File: Applied Course/4.Problem Solving/5.Problems on Trees/4.Print Level order traversal in spiral form Recursive Approach.py
import sys
class node: 

    def __init__(self, info): 
        self.info = info  
        self.left = None
        self.right = None 
    
def height(ptr):
    if ptr is None:
        return 0;
    
    lheight=height(ptr.left)
    rheight=height(ptr.right)
    if lheight>rheight:
        return 1+lheight
    else:
        return 1+rheight

def Levelorder(root,level,flag):
    if root is None:
        return
    if level==1:
        print(root.info)
    
    else: 
        if (flag is True):
            Levelorder(root.left,level-1,flag)
            Levelorder(root.right,level-1,flag)
        else:
            Levelorder(root.right,level-1,flag)
            Levelorder(root.left,level-1,flag)    

def printSpiral(root):
    h=height(root)
    flag=True
    for i in range(1,h+1):
        Levelorder(root,i,flag)
        flag=not flag
    
if __name__=='__main__':
    root=None
    root=node(10)
    root.left=node(20)
    root.right=node(30)
    root.left.left=node(40)
    root.left.right=node(50)
    root.right.left=node(60)
    root.right.right=node(70)
    printSpiral(root)
    

File: Applied Course/4.Problem Solving/5.Problems on Trees/4.Print Level order traversal in spiral form Recursive Approach.c
#include <stdbool.h> 
#include <stdio.h> 
#include <stdlib.h> 
  

struct node { 
    int data; 
    struct node* left; 
    struct node* right; 
}; 

/* Compute the "height" of a tree -- */
int height(struct node *root) 
{ 
    if(!root) 
        return 0; 
    else { 
        /* compute the height of each subtree */
        int lheight = height(root->left); 
        int rheight = height(root->right); 
        return (lheight >rheight)?(lheight+1): (rheight+1);
    } 
} 
  
//Create a new node
struct node* newNode(int data) 
{ 
    struct node *temp = (struct node*) 
        malloc(sizeof(struct node)); 
    temp->data = data; 
    temp->left = NULL; 
    temp->right = NULL; 
    return temp; 
} 

/* Print nodes at a given level */
void printGivenLevel(struct node* root, int level, int flag) 
{ 
    if(!root) 
        return; 
    if(level == 1) 
        printf("%d ", root->data); 
    else if (level > 1) { 
        if (flag) { 
            printGivenLevel(root->left, level - 1, flag); 
            printGivenLevel(root->right, level - 1, flag); 
        } 
        else { 
            printGivenLevel(root->right, level - 1, flag); 
            printGivenLevel(root->left, level - 1, flag); 
        } 
    } 
} 
    
/* Function to print spiral traversal of a tree*/
void printSpiral(struct node* root) 
{ 
    // get height of the tree
    int h = height(root); 
    int i; 
  
    /*flag -> Left to Right. If this variable is set, 
      then the given level is traversed from left to right. */
    int flag = 0; 
    for (i = 1; i <= h; i++) { 
        printGivenLevel(root, i, flag); 
  
        /*change flag value to traverse next level in opposite order*/
        flag = !flag; 
    } 
} 
  
int main() 
{ 
    struct node* root = newNode(10); 
    root->left = newNode(20); 
    root->right = newNode(30); 
    root->left->left = newNode(70); 
    root->left->right = newNode(60); 
    root->right->left = newNode(50); 
    root->right->right = newNode(40); 
    printf("Spiral Order traversal of binary tree is \n"); 
    printSpiral(root); 
  
    return 0; 
} 
